### PR TITLE
created a user model and implemented access to endpoints by their role

### PR DESCRIPTION
**Pull request by irynamatveieva**

### CODE REVIEW

File: src/main/java/com/example/bookstore/dto/user/UserLoginRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Size;
+import lombok.Data;
+
+@Data
+public class UserLoginRequestDto {
+    @NotBlank
+    @Size(min = 4, max = 50)
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/dto/user/UserLoginRequestDto.java]:
```suggestion
```

File: src/main/java/com/example/bookstore/dto/user/UserRegistrationRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import com.example.bookstore.validation.FieldMatch;
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Size;
+import lombok.Data;
+
+@Data
+@FieldMatch(firstField = "password", secondField = "repeatPassword")
+public class UserRegistrationRequestDto {
+    @NotBlank
+    @Size(min = 4, max = 50)
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/dto/user/UserRegistrationRequestDto.java]:
```suggestion
```

File: src/main/java/com/example/bookstore/dto/user/UserRegistrationRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package com.example.bookstore.dto.user;
+
+import com.example.bookstore.validation.Email;
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/dto/user/UserRegistrationRequestDto.java]: why you do not use` jakarta.validation.constraints.Email`?

File: src/main/java/com/example/bookstore/dto/user/UserRegistrationRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+public class UserRegistrationRequestDto {
+    @NotBlank
+    @Size(min = 4, max = 50)
+    @Email
+    private String email;
+    @NotBlank
+    @Size(min = 8, max = 100)
+    private String password;
+    @NotBlank
+    @Size(min = 8, max = 100)
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/dto/user/UserRegistrationRequestDto.java]:
```suggestion
```

File: src/main/java/com/example/bookstore/model/User.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+import java.util.stream.Collectors;
+import lombok.Data;
+import lombok.ToString;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+
+@Data
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/model/User.java]: better do not use @Data with @Entity classes
With @Data you must remember to add many `Exclude` annotations. password for example does not have it.
So, using @Data is a potential threat

File: src/main/java/com/example/bookstore/repository/book/BookRepository.java [hunk 6]
```diff
@@ -1,8 +1,25 @@
...
+    @Query("UPDATE Book b SET b.title = :title, b.author = :author, "
+            + "b.isbn = :isbn, b.price = :price, b.description = :description, "
+            + "b.coverImage = :coverImage WHERE b.id = :id AND b.isDeleted = FALSE")
+    void updateBookById(@Param(value = "id") Long id,
+                        @Param(value = "title") String title,
+                        @Param(value = "author") String author,
+                        @Param(value = "isbn") String isbn,
+                        @Param(value = "price") BigDecimal price,
+                        @Param(value = "description") String description,
+                        @Param(value = "coverImage") String coverImage);
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/repository/book/BookRepository.java]: we can use `save `for update

File: src/main/java/com/example/bookstore/repository/user/UserRepository.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package com.example.bookstore.repository.user;
+
+import com.example.bookstore.model.User;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+public interface UserRepository extends JpaRepository<User, Long> {
+    @Query("SELECT u FROM User u LEFT JOIN FETCH u.roles WHERE u.email = :email")
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/repository/user/UserRepository.java]:
```suggestion
    @EntityGraph(attributePaths = "roles")
```

File: src/main/java/com/example/bookstore/repository/user/UserRepository.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+    Optional<User> findUserByEmail(String email);
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/repository/user/UserRepository.java]:
```suggestion
    Optional<User> findByEmail(String email);
```

File: src/main/java/com/example/bookstore/service/book/BookService.java [hunk 9]
```diff
@@ -13,6 +13,8 @@ public interface BookService {
 
     List<BookDto> findAll(Pageable pageable);
 
+    void updateById(Long id, CreateBookRequestDto requestDto);
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/service/book/BookService.java]:
```suggestion
    BookDto updateById(Long id, CreateBookRequestDto requestDto);
```

File: src/main/java/com/example/bookstore/service/book/impl/BookServiceImpl.java [hunk 10]
```diff
@@ -28,6 +28,18 @@ public BookDto save(CreateBookRequestDto requestDto) {
...
+    @Override
+    public void updateById(Long id, CreateBookRequestDto requestDto) {
+        bookRepository.updateBookById(
+                id,
+                requestDto.getTitle(),
+                requestDto.getAuthor(),
+                requestDto.getIsbn(),
+                requestDto.getPrice(),
+                requestDto.getDescription(),
+                requestDto.getCoverImage());
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/service/book/impl/BookServiceImpl.java]:
```suggestion
        Book dbBook = repository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Can not found Book by id = " + id));
        Book book = bookMapper.toModel(bookDto);
        book.setId(id);
        return bookMapper.toDto(repository.save(book));
```

File: src/main/java/com/example/bookstore/service/user/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+            throws RegistrationException {
+        if (userRepository.findUserByEmail(requestDto.getEmail()).isPresent()) {
+            throw new RegistrationException("Unable to complete registration.");
+        }
+        User user = new User();
+        user.setEmail(requestDto.getEmail());
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+        user.setFirstName(requestDto.getFirstName());
+        user.setLastName(requestDto.getLastName());
+        user.setShippingAddress(requestDto.getShippingAddress());
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/service/user/impl/UserServiceImpl.java:37]: set Role USER

File: src/main/java/com/example/bookstore/validation/FieldMatchValidator.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    public boolean isValid(Object value, ConstraintValidatorContext context) {
+        try {
+            Field firstField = value.getClass().getDeclaredField(this.firstField);
+            Field secondField = value.getClass().getDeclaredField(this.secondField);
+            firstField.setAccessible(true);
+            secondField.setAccessible(true);
+            Object firstValue = firstField.get(value);
+            Object secondValue = secondField.get(value);
+            return (firstValue == null && secondValue == null)
+                    || (firstValue != null && firstValue.equals(secondValue));
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/validation/FieldMatchValidator.java]: Objects.equals

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/bookstore/dto/user/UserRegistrationRequestDto.java (hunk 3 above)
Comment (irynamatveieva) [reply src/main/java/com/example/bookstore/dto/user/UserRegistrationRequestDto.java]: forgot about the existence of this annotation. indeed, there is no need to create your own.

**Review by aaananas — APPROVED** [review]

**Pull request merged**